### ISSUE TITLE
Mi Band 3, Treadmill function doesn't log distance and thus exercise always end with "Distance is too short to be recorded".

### ISSUE THREAD

**Issue by ArmitageShanks** [Freeyourgadget/Gadgetbridge#1398]
#### Your wearable device is:
Mi Band 3. Firmware version 2.2.0.14

#### Your android version is:
Android 8.1

#### Your Gadgetbridge version is:
0.32.1

If you start the treadmill function on the Mi Band 3, it starts and shows info like time, HR, steps/min, steps and distance. All of them are displayed correctly except distance. It never changes from zero regardless of how long you have ran. Tried it numerous times on varying distances and it always shows zero, and then when you leave the function you always  get the ""Distance is too short to be recorded" error message. 

I'm guessing the Mi Band 3 doesnt have access to your step length/height from GB. 

Unfortunately I don't use the Mi Fit app so I dont know if the same issue exists there.

**Comment by ArmitageShanks** [comment]
It works for you? Hm, maybe they fixed it in the latest firmware then. I havent tried it yet since it said it wasnt tested on the wiki, but I'll give it a try tonight.

**Comment by ashimokawa** [comment]
@ArmitageShanks 
Please let me know the result. If there is some manual step setting or calibration in Mi Fit I would also like to know
(we never used any other app than Gadgetbridge and rely on outside information)

**Comment by ArmitageShanks** [comment]
I know updated to FW 2.3.0.2 but the problem persists, the treadmill option on the Band does not record distance, which results in that every time you use it you will get an error and it will not be saved.

**Comment by ee-gary** [comment]
I have the exact same issue. FW 2.3.0.2, Gadgetbridge 0.32.3.

On treadmill activity:
1. No distance displayed on band.
2. At completion of exercise, "Distance is too short to be recorded" warning

Mi Band 3 has only ever been connected to Gadgetbridge (never Mi Fit or other app). 

In "Language and region settings" section, units are set to 'Imperial'.
In "About you" section, height and weight are set (cm & kg).

**Comment by ashimokawa** [comment]
Damn it works for Nene even after factory reset :/

**Comment by ee-gary** [comment]
Under "Language and region settings" section, changed units to 'Metric'. Exited settings menu. Confirmed band was in metric mode. Still no distance recorded in treadmill mode.

Under "About you" section, changed height and weight settings slightly (cm & kg). Exited settings menu. Still no distance recorded in treadmill mode.

Are you aware of any other settings/variables that may be used by the treadmill function to display distance? It appears my GB + Band setup is not initializing something correctly, hence the distance = 0 issue.

**Comment by ArmitageShanks** [comment]
I have no idea why, but my Band 3 is now logging the distance. It's shown both when using the Treadmill function, but as a nice bonus, it's also seems like its logging the daily distance (and calorie burn, not sure how accurate either are though) under the info screen (the screen where you can see how many steps and what the battery status is, there are now two more sub screens there showing distance and calories). 

What has changed? I switched to another phone. Which is using Pie instead of Oreo. Also, this time when I set the Band up, I made sure that I had entered my data in GB before connecting the band for the first time. I also entered a random name for my profile (last time I hadnt bothered entering anything in this field the last time. 

This might have fixed the bug. 

Can anyone confirm adding a name*as well as age, sex, weight etc) when you set your Gadgetbridge up works for logging distance?

**Comment by ee-gary** [comment]
Based on your comment, I deleted the Miband 3 device from gadgetbridge (which deleted all my data).  I re-paired the device, entering name in my profile (weight/height already entered from before, but I had left name blank).

I now have the two new sub screens showing distance and calories!

**Comment by ee-gary** [comment]
When are 'about you' settings (height in cm, weight in kg) written to the MiBand 3? Is it only on initial setup?

Is there some reason 'Name/Alias' in 'Mi Band / Amazfit settings' needs to be set for distance and calorie sub screens to show up? Or was there a bug fix in gadgetbridge and this was fixed by re-pairing the miband 3 to the latest version of gadgetbridge (0.32.4)?

For my future reference, is there a way to retain my steps if I need to delete & re-pair my miband 3?

Gary

**Comment by ArmitageShanks** [comment]
Nice to hear this "fix" makes it work then!

I assumed that it was written on the band on the first setup since you get the "About you" screen popping up on the first setup, and I assumed that not putting a name in didnt make the settings stick. But when you mention it, maybe it gets triggered when you put name in, and not at setup? 

But either way, I consider this problem fixed for me. Not sure whether to close the issue or not though, since there obviously is an issue somewhere.

**Comment by Uatschitchun** [comment]
The 2 new screens you're talking about here are on the Band, not in the app, correct? Is it possible to get this info in the app, too?

**Comment by ArmitageShanks** [comment]
Yep, only on the band. The only place I've even seen it mentioned in the app is in the settings for Mi Band 2 and what items to display on the band.

Not sure if the data is particularly useful though since the distance is always off. If you're running it's about half the distance you actually ran, and if you're walking it's usually 1/3 off the actual distance.

**Comment by ee-gary** [comment]
Presumably, the stride length is based off of 'Gender' and 'Height in cm' in the Gadgetbridge 'About you' settings. For example, google search indicates Fitbit uses 0.415x height for men and 0.413x height for women.

1. I'm not sure what formula the MiBand3 uses. I've tried to back-calculate based on reported steps, miles, and height setting but it's not consistent. Presumably, if distance is under-reported, we can increase the height setting in gadgetbridge to calibrate.
2. I don't know what triggers the 'About you' settings to be sent to the band. Any time they are updated? Only when setting up a new band? Perhaps one of the devs could comment...
